packages/expo-updates/android/src/main/java/expo/modules/updates/statemachine/UpdatesStateMachine.kt

```diff
@@ -0,0 +1,199 @@
+package expo.modules.updates.statemachine
+
+import android.content.Context
+import com.facebook.react.bridge.Arguments
+import com.facebook.react.bridge.WritableMap
+import expo.modules.updates.logging.UpdatesLogger
+import org.json.JSONObject
+
+/**
+ * The Updates state machine class. There should be only one instance of this class
+ * in a production app, instantiated as a property of UpdatesController.
+ */
+class UpdatesStateMachine constructor(
+  androidContext: Context,
+  val changeEventSender: UpdatesStateChangeEventSender
+) {
+
+  private val logger = UpdatesLogger(androidContext)
+
+  /**
+   * The current state
+   */
+  var state: UpdatesStateValue = UpdatesStateValue.Idle
+
+  /**
+   * The context
+   */
+  var context: UpdatesStateContext = UpdatesStateContext()
+
+  /**
+   Called after the app restarts (reloadAsync()) to reset the machine to its
+   starting state.
+   */
+  fun reset() {
+    state = UpdatesStateValue.Idle
+    context = UpdatesStateContext()
+    logger.info("Updates state change: reset, context = ${context.json}")
+    sendChangeEventToJS()
+  }
+
+  /**
+   Called by LoaderTask delegate methods in UpdatesController during the initial
+   background check for updates, and called by checkForUpdateAsync(), fetchUpdateAsync(), and reloadAsync().
+   */
+  fun processEvent(event: UpdatesStateEvent) {
+    if (transition(event)) {
+      context = reduceContext(context, event)
+      logger.info("Updates state change: ${event.type}, context = ${context.json}")
+      // Send change event
+      sendChangeEventToJS(event)
+    }
+  }
+
+  /**
+   Make sure the state transition is allowed, and then update the state.
+   */
+  private fun transition(event: UpdatesStateEvent): Boolean {
+    val allowedEvents: Set<UpdatesStateEventType> = updatesStateAllowedEvents[state] ?: setOf()
+    if (!allowedEvents.contains(event.type)) {
+      // Optionally put an assert here when testing to catch bad state transitions in E2E tests
+      return false
+    }
+    state = updatesStateTransitions[event.type] ?: UpdatesStateValue.Idle
+    return true
+  }
+
+  /**
+   * If a state change event is passed in, the JS sender
+   * is called with just the fields and values that changed.
+   * During a reset, this method is called with no event passed in,
+   * and then all the fields and the entire context are passed to the JS sender.
+   */
+  private fun sendChangeEventToJS(event: UpdatesStateEvent? = null) {
+    changeEventSender.sendUpdateStateChangeEventToBridge(
+      event?.type ?: UpdatesStateEventType.Restart,
+      context.json
```

Comment:
instead of passing JSON and then calling back into this class to convert to writable map, just pass the value of `paramsForJSEvent`. Even better, add a `toWritableMap` method to `UpdatesStateContext` so that the hard to follow `field.startsWith("is")` and `field.endsWith("Manifest")` are not necessary.

Reply:
I will try that -- however I originally did things this way because I found that attempting to create WritableMaps didn't work unless the bridge was available, so things broke in unit tests. I'll have another look to see if it is possible to go directly to WritableMap.